src/dataqueue/queue.cc

```diff
@@ -0,0 +1,1221 @@
+               ->NewInstance(env->context())
+               .ToLocal(&obj)) {
+        return BaseObjectPtr<Wrap>();
+      }
+
+      return MakeBaseObject<Wrap>(env, obj, fd, start, end);
+    }
+
+    Wrap(Environment* env,
+         Local<Object> obj,
+         int fd,
+         uint64_t start,
+         v8::Maybe<uint64_t> end)
+        : BaseObject(env, obj),
+          inner_(std::make_unique<FdEntry>(env, fd, start, end)) {
+      MakeWeak();
+    }
+
+    std::unique_ptr<DataQueue::Entry> detach() { return std::move(inner_); }
+
+    bool isDetached() const { return inner_ == nullptr; }
+
+    void MemoryInfo(MemoryTracker* tracker) const override {
+      tracker->TrackField("entry", inner_);
+    }
+    SET_MEMORY_INFO_NAME(FdEntry::Wrap)
+    SET_SELF_SIZE(Wrap)
+
+   private:
+    std::unique_ptr<FdEntry> inner_;
+  };
+
+ private:
+  Environment* env_;
+  int fd_;
+  uint64_t start_ = 0;
+  uint64_t end_ = 0;
+  uv_stat_t stat_;
+  uv_fs_t req;
+  BaseObjectPtr<fs::FileHandle> maybe_file_handle_;
+
+  int GetStat(uv_stat_t* stat) {
+    int err = uv_fs_fstat(env_->event_loop(), &req, fd_, nullptr);
+    *stat = req.statbuf;
+    return err;
+  }
+
+  class Reader : public DataQueue::Reader {
+   public:
+    explicit Reader(FdEntry* entry)
+        : entry_(entry), offset_(entry->start_), end_(entry_->end_) {}
+
+    int Pull(Next next,
+             int options,
+             DataQueue::Vec* data,
+             size_t count,
+             size_t max_count_hint = bob::kMaxCountHint) override {
+      // TODO(@jasnell): For now, we're going to ignore data and count.
+      // Later, we can support these to allow the caller to allocate the
+      // buffers we read into. To keep things easier for now, we're going
+      // to read into a pre-allocated buffer.
+      if (ended_ || offset_ == end_) {
+        std::move(next)(bob::STATUS_EOS, nullptr, 0, [](uint64_t) {});
+        return bob::STATUS_EOS;
+      }
+      // offset_ should always be less than end_ here
+      CHECK_LT(offset_, end_);
+      new PendingRead(this, std::move(next));
+      return bob::STATUS_WAIT;
+    }
+
+    SET_NO_MEMORY_INFO()
+    SET_MEMORY_INFO_NAME(FdEntry::Reader)
+    SET_SELF_SIZE(Reader)
+
+   private:
+    FdEntry* entry_;
+    bool ended_ = false;
+    uint64_t offset_;
+    uint64_t end_;
+
+    struct PendingRead {
+      static constexpr uint64_t DEFAULT_BUFFER_SIZE = 4096;
+      Reader* reader;
+      Next next;
+      uv_fs_t req_;
+      uv_buf_t uvbuf;
+
+      PendingRead(Reader* reader, Next next)
+          : reader(reader),
+            next(std::move(next)),
+            uvbuf(reader->entry_->env()->allocate_managed_buffer(std::min(
+                DEFAULT_BUFFER_SIZE, reader->end_ - reader->offset_))) {
+        req_.data = this;
+        uv_fs_fstat(reader->entry_->env()->event_loop(),
+                    &req_,
+                    reader->entry_->fd_,
+                    &PendingRead::OnStat);
```

Comment:
Is there  a good reason not to use `ReqWrap` (or `FSReqBase` or any other subclass of `ReqWrap`)? If there is, you’ll need to do the things that `ReqWrap` would otherwise do manually, i.e. async tracking, updating the waiting request counter, etc. (in particular that last item should be blocking this PR)